search/double_linear_search.ts

```diff
@@ -0,0 +1,28 @@
+/**
+ * @function doubleLinearSearch
+ * @description double linear search is an improvement of linear
```

Comment:
That's the thing: It isn't. Sure you have half as many loop iterations, but now each of those iterations is twice as expensive (roughly). The only case where it could be an improvement is if your elements are somehow less likely to be in the middle and more likely to be at the end(s).

Either way I don't see much of a value in this vs. just linear search so far. It's just overcomplicated for no tangible benefit.

Reply:
Thank you. Would you please explain what you mean when you say that each iteration is twice as expensive? Because it is just one iteration but we are using two pointers. Either way, when you talk about expensive, is it in terms of space? Please clarify, thank you.

Reply:
I'm talking about time complexity. Each iteration does (roughly) twice the work. You change two indices, you index two times, you do two comparisons.

Reply:
> I'm talking about time complexity. Each iteration does (roughly) twice the work. You change two indices, you index two times, you do two comparisons.

I agree with you, but if you come to think of it, we don't have to search half of the list, the only worst case of this algorithm is when the target element is very close to the mid index and I see where your point. I also agree with the fact that each iteration performs approximately twice the work due to managing two indices and comparisons, however, this doesn't change the fact that the overall time complexity remains O(n). Please help me understand better if I'm wrong, but I usually understand that we focus on the growth rate of the algorithm relative to the input size and not the exact number of operations per iteration.

Reply:
The time complexity is $O(n)$, just like linear search. Hence why I think this addition isn't well motivated, it's just an overcomplicated "double" linear search.

Reply:
> The time complexity is O ( n ) , just like linear search. Hence why I think this addition isn't well motivated, it's just an overcomplicated "double" linear search.

Great, I’m glad we’re on the same page! Now, let’s consider an array like `[10, 1, 13, 22, 12, 9, 54, 23, 12, 34, 15, 18] `with a target value of `15`. Don’t you think a double linear search could be faster? In certain scenarios, I believe double linear search can outperform a regular linear search. For instance, in this example, where the target is near the end of the array, a regular linear search would need to traverse through all the elements to find it. On the other hand, a double linear search, which processes from both ends simultaneously, could find the target with significantly fewer comparisons.

Reply:
> I already mentioned that the best cases differ, but the average and worst cases do not. It's really just linear search but visiting the elements in a different order. I don't think this is well motivated enough. Otherwise every variant of linear search, every permutation of the indices, would have a place in this project, which they do not. You can imagine infinitely many of these trivial variations of linear search (say, "even-odd" linear search, "triple" linear search, "double" linear search but you start from the middle, etc etc) and they don't add much vs a stock linear search.

Understood. Thank you for the eye opener.